rogg_granule.c

```diff
@@ -0,0 +1,160 @@
+   LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+/* ogg vorbis logical stream granule adjustment using the rogg library */
+
+/* compile with
+   gcc -O2 -g -Wall -I. -o rogg_granule rogg.c rogg_granule.c
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <rogg.h>
+
+int granule_adjust = 0;
+
+void print_usage(FILE *out, char *name)
+{
+  fprintf(stderr, "Script for editing ogg headers\n");
+  fprintf(stderr, "%s [-g n] <file1.ogg> [<file2.ogg>...]\n",
+	name);
+  fprintf(stderr,
+		  "    -g n        change the granule position of a logical vorbis stream by n\n"
+		  "                (can be positive or negative)\n"
+		  "\n");
+}
+
+void parse_args(int *argc, char *argv[])
+{
+  int arg = 1;
+  int shift;
+
+  while (arg < *argc) {
+    shift = 0;
+    if (argv[arg][0] == '-') {
+      switch (argv[arg][1]) {
+	case 'g':
+	  /* read granule adjustment from the next arg */
+	  if (sscanf(argv[arg+1], "%i", &granule_adjust) != 1) {
+	    fprintf(stderr, "Could not parse granule adjustment '%s'.\n", argv[arg+1]);
+	    granule_adjust = 0;
+	  } else {
+	    fprintf(stdout, "Adjusting granule position by %i\n", granule_adjust);
+	  }
+	  shift = 2;
+	  break;
+      }
+    }
+    if (shift) {
+      int left = *argc - arg - shift;
+      if (left < 0) {
+	fprintf(stderr, "Internal error parsing argument '%s'.\n", argv[arg]);
+	exit(1);
+      }
+      memmove(&argv[arg], &argv[arg+shift], left*sizeof(*argv));
+      *argc -= shift;
+    } else {
+      arg++;
+    }
+  }
+}
+
+int main(int argc, char *argv[])
+{
+  int f, i;
+  unsigned char *p, *q, *o, *e;
+  struct stat s;
+  rogg_page_header header;
+  uint64_t granulepos;
+
+  parse_args(&argc, argv);
+  if (argc < 2) {
+    print_usage(stderr, argv[0]);
+    exit(1);
+  }
+
+  for (i = 1; i < argc; i++) {
+    f = open(argv[i], O_RDWR);
+    if (f < 0) {
+	fprintf(stderr, "couldn't open '%s'\n", argv[i]);
+	continue;
+    }
+    if (fstat(f, &s) < 0) {
+	fprintf(stderr, "couldn't stat '%s'\n", argv[i]);
+	close(f);
+	continue;
+    }
+    p = mmap(0, s.st_size, PROT_READ|PROT_WRITE,
+	MAP_SHARED, f, 0);
+    if (p == MAP_FAILED) {
+	fprintf(stderr, "couldn't mmap '%s'\n", argv[i]);
+	close(f);
+	continue;
+    }
+    fprintf(stdout, "Checking Ogg file '%s'\n", argv[i]);
+    e = p + s.st_size; /* pointer to the end of the file */
+    q = rogg_scan(p, s.st_size); /* scan for an Ogg page */
+    if (q == NULL) {
+	fprintf(stdout, "couldn't find ogg data!\n");
+    } else {
+      if (q > p) {
+	fprintf(stdout, "Skipped %d garbage bytes at the start\n", (int)(q-p));
+      }
+      while (q < e) {
+	o = rogg_scan(q, e-q); /* find the next Ogg page */
```

Comment:
I'm not sure whether that would actually be valid. At first, I believed you were right, but after reading the [Vorbis mapping spec](https://www.xiph.org/vorbis/doc/Vorbis_I_spec.html#x1-132000A.2l) again, I'm not so sure anymore:

> The granule (PCM) position of the first page need not indicate that the stream started at position zero. Although the granule position belongs to the last completed packet on the page and a valid granule position must be positive, by inference it may indicate that the PCM position of the beginning of audio is positive or negative.

The question is, whether 0 would be considered positive in the paragraph. I think it should be considered positive, making a page granulepos 0 at the start of the stream invalid. I'm not exactly sure of the implications of lapping inside vorbis, but I think, the encoder/muxer should be able to just drop the packet which would have caused the granulepos of the first data page to be 0.

When applied to other codecs, it depends on their respective mapping specification whether data pages with granulepos 0 are allowed.
I had intentionally kept the logic codec-agnostic so far, but if I need to skip header packets/pages, I do not think that would possible any more (an Ogg page does not have a flag for "this is a stream header page").
Are there any other known Ogg codecs where adjusting the granulepos the way rogg_granule does it even makes sense? From a quick glance at [Speex,](http://www.speex.org/docs/manual/speex-manual/node8.html), it uses a similar granulepos mapping as vorbis, but does not specify at all what would happen with offset granulepos.
Opus works differently (with respect to initial offset). Theora also works differently. No idea about FLAC, Daala, OGM, or any of the metadata/subtitles codecs/streams.

Maybe just providing another command-line argument to skip a given number of initial packets (defaulting to 3) (and error-ing out if these do not terminate on a page boundary) would be a reasonable compromise here.

In any case, the file generated by rogg_granule may violate the following paragraph of the Vorbis spec:

> In both of these cases in which the initial audio PCM starting offset is nonzero, the second finished audio packet must flush the page on which it appears and the third packet begin a fresh page. This allows the decoder to always be able to perform PCM position adjustments before needing to return any PCM data from synthesis, resulting in correct positioning information without any aditional seeking logic.
> **Note:** Failure to do so should, at worst, cause a decoder implementation to return incorrect positioning information for seeking operations at the very beginning of the stream.

Without the implementation of a proper remuxer (which I have no intention to implement for rogg_granule), this cannot be done properly.
Using libvorbisfile in a streaming context (i.e. not ever seeking the Vorbis stream) handles such files fine though.